cmd/gke-serviceaccounts-initializer/main.go

```diff
@@ -0,0 +1,223 @@
+					log.Printf("error saving pod/%s: %+v", pod.GetName(), err)
+				} else {
+					log.Printf("initialized pod/%s", pod.GetName())
+				}
+			},
+		},
+	)
+
+	stop := make(chan struct{})
+	go controller.Run(stop)
+
+	signalChan := make(chan os.Signal, 1)
+	signal.Notify(signalChan, syscall.SIGINT, syscall.SIGTERM)
+	<-signalChan
+
+	log.Println("Shutdown signal received, exiting...")
+	close(stop)
+}
+
+// needsInitialization determines if the pod is required to be initialized
+// currently by this initializer.
+func needsInitialization(pod *corev1.Pod) bool {
+	initializers := pod.ObjectMeta.GetInitializers()
+	return initializers != nil &&
+		len(initializers.Pending) > 0 &&
+		initializers.Pending[0].Name == initializerName
+}
+
+// removeSelfPendingInitializer removes the first element from pending
+// initializers list of in-memory pod value.
+func removeSelfPendingInitializer(pod *corev1.Pod) {
+	if pod.ObjectMeta.GetInitializers() == nil {
+		return
+	}
+	pendingInitializers := pod.ObjectMeta.GetInitializers().Pending
+	if len(pendingInitializers) == 1 {
+		pod.ObjectMeta.Initializers.Pending = nil
+	} else if len(pendingInitializers) > 1 {
+		pod.ObjectMeta.Initializers.Pending = append(
+			pendingInitializers[:0], pendingInitializers[1:]...)
+	}
+}
+
+// patchPod saves the pod to the API using a strategic 2-way JSON merge patch.
+func patchPod(origPod, newPod *corev1.Pod, clientset *kubernetes.Clientset) error {
+	origData, err := json.Marshal(origPod)
+	if err != nil {
+		return fmt.Errorf("failed to marshal original pod: %+v", err)
+	}
+
+	newData, err := json.Marshal(newPod)
+	if err != nil {
+		return fmt.Errorf("failed to marshal modified pod: %+v", err)
+	}
+
+	patch, err := strategicpatch.CreateTwoWayMergePatch(origData, newData, corev1.Pod{})
+	if err != nil {
+		return fmt.Errorf("failed to create 2-way merge patch: %+v", err)
+	}
+
+	if _, err = clientset.CoreV1().Pods(origPod.GetNamespace()).Patch(
+		origPod.GetName(), types.StrategicMergePatchType, patch); err != nil {
+		return fmt.Errorf("failed to patch pod/%s: %+v", origPod.GetName(), err)
+	}
+	return nil
+}
+
+// modifyPodSpec makes modifications to in-memory pod value to inject the
+// service account. Returns whether any modifications have been made.
+func modifyPodSpec(pod *corev1.Pod) bool {
+	if pod == nil || pod.ObjectMeta.Annotations == nil {
+		return false
+	}
+	serviceAccountName, ok := pod.ObjectMeta.Annotations[annotation]
+	if !ok {
+		return false
+	}
+
+	volName := fmt.Sprintf("gcp-%s", serviceAccountName)
+	mountPath := path.Join(secretMountPath, serviceAccountName)
+	keyPath := path.Join(mountPath, serviceAccountFile)
+
+	pod.Spec.Volumes = append(pod.Spec.Volumes,
+		corev1.Volume{
+			Name: volName,
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName: serviceAccountName,
+					Items: []corev1.KeyToPath{{
+						Key:  "key.json",
+						Path: "key.json",
```

Comment:
Is `key.json` a typo?

Reply:
nope, this works fine today. this makes sure we only import key.json (and nothing else) from that Secret, and put that on key.json.